src/data/bitwise.rs

```diff
@@ -65,11 +67,21 @@ impl Shl for Number {
 
     #[inline]
     fn shl(self, rhs: Self) -> Self::Output {
-        assert!(self.layout().is_integer(), "bit shifting float number");
-        self.to_u1024_bytes()
-            .shl(u16::try_from(rhs).expect("attempt to bitshift left for more than 2^16 bits")
-                as usize)
-            .into()
+        let layout = self.layout();
+        assert!(layout.is_integer(), "bit shifting float number");
+        let rhs = u16::try_from(rhs).expect("attempt to bitshift lhs for more than 2^16 bits");
+        match layout.is_signed_int() {
+            true => {
+                Number::from(self.to_i1024_bytes().checked_shl(rhs as u32).unwrap_or(i1024::ZERO))
+                    .reshaped(layout, true)
+                    .expect("discarding excess bits")
```

Comment:
Here and below: we must have no exceptions on code execution for AluVM opcodes. It is ok to have exception if we try to SHL/SHR on floats, since there is no opcode for float bit shift - meaning it will be an error in the AluVM code itself and not the AluVM bytecode. But throwing on bit discards will result in AluVM failures which are unacceptable. We must set st0 to false in these case and continue execution!

Reply:
This will never throw and I am just unwrapping the value here. `reshaped` method always returns some if `wrap` flag is `true`

https://github.com/internet2-org/rust-aluvm/blob/5da753a529fc8fa0383858c53b8c9c8ea188e58f/src/data/number.rs#L841

Reply:
Maybe we can improve the except message than, saying that reshape function is broken

Reply:
I think it is my bad to unwrap the value so easily. I am refactoring the code

Reply:
```rust
let mut n = match layout.is_signed_int() {
            true => {
                Number::from(self.to_i1024_bytes().checked_shr(rhs as u32).unwrap_or(i1024::ZERO))
            }
            false => {
                Number::from(self.to_u1024_bytes().checked_shr(rhs as u32).unwrap_or(u1024::ZERO))
            }
        };
n.reshape(layout);
n
```

Reply:
done